### ISSUE TITLE
[Feature request]: Expose a histogram of sstables reads per CQL read

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#11981]
*Description*
It would be very helpful to expose a histogram (at least average, P95, P99, MAX) for the number of ssstables read in the context of a single CQL read.

This is very helpful when we want to estimate the state of sstables set on a node.
For instance, if we see that we need to read too many sstables for a single read it may indicate that compaction is not very efficient.

This data is available via `nodetool cfhostograms` but it's much more practical to have it in Prometheus metrics.

**Comment by isburmistrov** [comment]
Having such a metric on Scylla's Prometheus stack would be really really helpful. 

First, `nodetool cfhostograms` are currently broken, see https://supportscylladb.zendesk.com/hc/en-us/requests/2936.

Second, for a managed Scylla customers requesting it every single time via tickets is inconvenient.

And third and most important, Scylla's dashboard may add a huge power to this. Imagine having sstable reads histogram per shard - such a metric may indicate that some shards are more impacted than the others, due to e.g. some skew in data.

**Comment by tzach** [comment]
@vladzcloudius I suspect mixing all CQL reads into one histogram will make it less valuable.
Is it better to tag one query and generate a histogram?
If this is the case, a mechanism like tracing might be better.

**Comment by dorlaor** [comment]
If we collect this data anyway, it is supposed to be an easy decision, the nodetool
is an operator mechanism and many times isn't available to the end user.

**Comment by amnonh** [comment]
I disagree with the solution. It's a very high cardinality, time-series databases (i.e., Prometheus) are the wrong solution.
If you try to collect per-table metrics, you end up with millions of metrics.
I think the right way forward is with CQL API - virtual table or call procedures.

**Comment by dorlaor** [comment]
On Tue, Nov 15, 2022 at 9:45 AM Amnon Heiman ***@***.***>
wrote:

> I disagree with the solution. It's a very high cardinality, time-series
> databases (i.e., Prometheus) are the wrong solution.
> If you try to collect per-table metrics, you end up with millions of
> metrics.
> I think the right way forward is with CQL API - virtual table or call
> procedures.
>

Virtual table is the best api, the caveat is that we'll lose lots of the
time aspect but if
we show the histogram since the last call and other x-time history, it
should be ok.
At the end of the day, 99.99% of users shouldn't care and this is an
operator issue

**Comment by mykaul** [comment]
> If we collect this data anyway, it is supposed to be an easy decision, the nodetool is an operator mechanism and many times isn't available to the end user.

I don't see how this metric is relevant to the end user.

**Comment by tzach** [comment]
Histogram for what metric? Is it per table or per query?
Both cardinalities can explode.

There are already mechanisms to select and trace *specific* query
It makes more sense to enhance them to collect info, regardless of whether this info ends up in the CQL table or the monitoring stack.

**Comment by michoecho** [comment]
> The question, do you need a minute by minute 45d history for each table on each shard?

Not everything has to be minute by minute. Maybe the amount of data can be kept in check by scraping the expensive metrics with a longer interval? Or metrics can be multiplexed: on each scrape call, let Scylla sample just one random table for its per-table metrics, and send just the results for that one table to Prometheus. This way you get per-table metrics downsampled by (1/number of tables) on average, avoiding a cardinality explosion while retaining good resolution for clusters with a reasonable number of tables.

**Comment by isburmistrov** [comment]
@mykaul 

> I don't see how this metric is relevant to the end user.

also @tzach w.r.t. tracing comments

See, I'm the end user and I need this badly :) Why? 

Because currently it's impossible to understand if the current choice of compaction strategy is good or bad. What I want to be able to answer is this question: "what contributes to my 99% latency being XXms?".

There could be multiple things, like:

- too big concurrency
- many cache misses
- many sstables being read for 1% of requests

First two can be answered using the dashboard we have, the third one can't.  Can it be that due to the specifics of the workload we need to switch from ICS to LCS? Or e.g. tune amplification goal in ICS? Or something else?
These considerations will be completely different depending on how much SSTables we read per single request. 

For instance, if p99 for sstables per read is 1 - it says with confident that our compaction strategy perfectly matches our workload. If it's 2 - probably too. But if it's 10, then we should think if something needs to be changed.

Any idea how this analysis can be done without having this metric exposed? Tracing is not the answer - single request can touch 10 sstables, it say nothing about 1% of requests. nodetool is not the answer too due to the reasons mentioned above. 
Any other ways? Happy to learn if I'm missing something :)

**Comment by tzach** [comment]
@isburmistrov I do not doubt it would be useful - when required.
But, the reporting cost is too high. We need a way to limit the number of metrics, exposing just a minimal relevant subset.

TRACING is an example of such a way; one can only trace slow queries or only trace 1 or 1000,000 requests.

**Comment by isburmistrov** [comment]
@tzach 

Ah sorry, you mean enabling probabilistic tracing and then looking into tracing table - this can work yes!

By the way, speaking about high cardinality. 
if Scylla can export tracing data using Open Telemetry, one can setup things like https://www.honeycomb.io/ and have an endless possibilities of digging / debugging.

**Comment by mykaul** [comment]
> @mykaul
> 
> > I don't see how this metric is relevant to the end user.
> 
> also @tzach w.r.t. tracing comments
> 
> See, I'm the end user and I need this badly :) Why?

Thanks for the insight! See below for questions I still have.

> 
> Because currently it's impossible to understand if the current choice of compaction strategy is good or bad. What I want to be able to answer is this question: "what contributes to my 99% latency being XXms?".
> 
> There could be multiple things, like:
> 
> * too big concurrency
> * many cache misses
> * many sstables being read for 1% of requests

Let's assume you have this metric, and it indeed tells you that you have too many sstables read. What, as an end user, will you do?

I'd like to ensure we have actionable alerts and metrics - those that you could easily deduce from what's the corrective action should be. It may be simple or more complex (reducing concurrency is an example - should be fairly easy to implement, but may not satisfy the business requirement, and could go all the way to we need stronger cluster, etc.)

> 
> First two can be answered using the dashboard we have, the third one can't. Can it be that due to the specifics of the workload we need to switch from ICS to LCS? Or e.g. tune amplification goal in ICS? Or something else? These considerations will be completely different depending on how much SSTables we read per single request.
> 
> For instance, if p99 for sstables per read is 1 - it says with confident that our compaction strategy perfectly matches our workload. If it's 2 - probably too. But if it's 10, then we should think if something needs to be changed.

OK, so the answer is lets improve the compaction strategy. It's unfortunate that end user has to still play with different metrics to  optimize that - I think it's an area that needs improvement.

> 
> Any idea how this analysis can be done without having this metric exposed? Tracing is not the answer - single request can touch 10 sstables, it say nothing about 1% of requests. nodetool is not the answer too due to the reasons mentioned above. Any other ways? Happy to learn if I'm missing something :)

I remember some work on open telemetry - can't find it right now.
Thanks again for the valuable feedback!

**Comment by isburmistrov** [comment]
> OK, so the answer is lets improve the compaction strategy. It's unfortunate that end user has to still play with different metrics to optimize that - I think it's an area that needs improvement.

Of course, the ultimate goal would be self-tuning compaction strategy right? :) 
But we're not there yet and have long time to go to that state. In the meantime some form of the observability is needed to make the right call.

Don't mind the exact implementation details - it may not be Prom neither it should be for per-minute granularity, just need this observability to be in place.

**Comment by vladzcloudius** [comment]
> Histogram for what metric? Is it per table or per query? Both cardinalities can explode.
> 
> There are already mechanisms to select and trace _specific_ query. 
> It makes more sense to enhance them to collect info, regardless of whether this info ends up in the CQL table or the monitoring stack.

Not sure what you are talking about, @tzach.
And no, this request is for a metric today collected on a per-table level - not per-query.

**Comment by vladzcloudius** [comment]
> I'm unassigning myself until there's a decision of what needs to be done and where. From what we've seen in the past a naive approach of adding per shard/table metric simply doesn't not work. Sure if you have a tiny cluster and a few tables it would be ok.
> 
> when there's a decision what to do, re-assign me if needed

`nodetool cfhistograms` doesn't expose a per-shard info. 
And having a node-level P99 is already plenty useful.
And it won't explode, @amnonh.

**Comment by denesb** [comment]
> nodetool cfhistograms doesn't expose a per-shard info.

Now that we have our native nodetool, nothing stops us from making it shard-aware.
